### PR TITLE
adding `print_ard_conditions()`

### LINKED ISSUE #118: Add function to print errors and warnings
Perhaps something like this:

```r
ard <- 
  mtcars |> 
  cards::ard_continuous(
    variables = c("mpg", "hp"),
    statistics = ~list(
      mean = \(x) mean(x),
      mean_warn = \(x) {warning("This is your warning!"); mean(x)},
      errr = \(x) stop("That's it! STOP!")
    )
  )
ard

print_messages <- function(x) {
  ard_warning <- x |> dplyr::filter(cards:::map_lgl(.data$warning, Negate(is.null)))
  ard_error <- x |> dplyr::filter(cards:::map_lgl(.data$error, Negate(is.null)))
  
  
  # this section needs to be updated to: 
  #  1. add the variable level if populated
  #  2. add the grouping variables/levels if populated
  cli::cli_bullets(c(
    "The following warnings were observed:",
    "!" = "Variable {.var {ard_warning$variable[[1]]}} for statistic {.val {ard_warning$stat_name[[1]]}}: {cli::col_yellow(ard_warning$warning[[1]])}",
    "!" = "Variable {.var {ard_warning$variable[[2]]}} for statistic {.val {ard_warning$stat_name[[2]]}}: {cli::col_yellow(ard_warning$warning[[2]])}"
  ))
  
  cli::cli_bullets(c(
    "The following errors occured:",
    "x" = "Variable {.var {ard_error$variable[[1]]}} for statistic {.val {ard_error$stat_name[[1]]}}: {cli::col_red(ard_error$error[[1]])}",
    "x" = "Variable {.var {ard_error$variable[[2]]}} for statistic {.val {ard_error$stat_name[[2]]}}: {cli::col_red(ard_error$error[[2]])}"
  ))
}

print_messages(ard)
```

image

For ARDs like `ard_ttest()` we can have one error for every statistic returned. In these cases, the messages should be grouped by the groups, variables, and error message. We can list all the statistics the error applies applies to.

### PR DESCRIPTION
**Pull request by ddsjoberg**
**What changes are proposed in this pull request?**
Adds the `print_ard_conditions()` function

**Reference GitHub issue associated with pull request.** _e.g., 'closes #1'_
closes #118 

image

--------------------------------------------------------------------------------

Reviewer Checklist (if item does not apply, mark is as complete)

- [ ] Ensure all package dependencies are installed: `devtools::install_dev_deps()`
- [ ] PR branch has pulled the most recent updates from master branch: `usethis::pr_merge_main()`
- [ ] If a bug was fixed, a unit test was added.
- [ ] Run `pkgdown::build_site()`. Check the R console for errors, and review the rendered website.
- [ ] Code coverage is suitable for any new functions/features: `devtools::test_coverage()`
- [ ] `usethis::use_spell_check()` runs with no spelling errors in documentation

When the branch is ready to be merged:
- [ ] Update `NEWS.md` with the changes from this pull request under the heading "`# cards (development version)`". If there is an issue associated with the pull request, reference it in parentheses at the end update (see `NEWS.md` for examples).
- [ ] Increment the version number using `usethis::use_version(which = "dev")`
- [ ] Run `usethis::use_spell_check()` again
- [ ] Approve Pull Request
- [ ] Merge the PR. Please use "Squash and merge".

### CODE REVIEW

File: tests/testthat/test-print_ard_conditions.R [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        err_fn = \(x) stop("'tis an error")
+      )
+    ) |>
+      print_ard_conditions()
+  )
+
+  # expected messaging when the same error appears for all stats (consolidated correctly)
+  expect_snapshot(
+    ard_continuous(ADSL, variables = AGE) |>
+      dplyr::mutate(error = list("repeated error")) |>
```
Comment (bzkrouse) [inline tests/testthat/test-print_ard_conditions.R:40]: nice!!

**Review by bzkrouse — CHANGES_REQUESTED** [review]
love this! Just flagging the check_class call!

File: R/print_ard_conditions.R [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+#'   variables = AGE,
+#'   statistics = ~list(
+#'     mean = \(x) mean(x),
+#'     mean_warning = \(x) {warning("warn1"); warning("warn2"); mean(x)},
+#'     err_fn = \(x) stop("'tis an error")
+#'   )
+#' ) |>
+#'   print_ard_conditions()
+print_ard_conditions <- function(x) {
+  check_class(x, class = "card", call = env)
```
[the remark this answers is not in the document]
Comment (ddsjoberg) [reply R/print_ard_conditions.R]: Oooops! That was leftover from something i deleted 😱 Good catch!!

**Comment by ddsjoberg** [comment]
Thanks for reviewing @bzkrouse ! Forgot to mention that when we can resolve #128 and #129, we'll make updates to this function. But I think it should be fine to address those at a later time.

**Review by bzkrouse — APPROVED** [review]
🎉

**Pull request merged**